pkg/controller/api.go

```diff
@@ -0,0 +1,101 @@
+/*
+ Copyright 2021 The Crossplane Authors.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	xpresource "github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrl "sigs.k8s.io/controller-runtime/pkg/manager"
+
+	"github.com/crossplane-contrib/terrajet/pkg/resource"
+	"github.com/crossplane-contrib/terrajet/pkg/terraform"
+)
+
+// APISecretClient is a client for getting k8s secrets
+type APISecretClient struct {
+	kube client.Client
+}
+
+// GetSecretData gets and returns data for the referenced secret
+func (a *APISecretClient) GetSecretData(ctx context.Context, ref *xpv1.SecretReference) (map[string][]byte, error) {
+	secret := &v1.Secret{}
+	if err := a.kube.Get(ctx, types.NamespacedName{Namespace: ref.Namespace, Name: ref.Name}, secret); err != nil {
+		return nil, err
+	}
+	return secret.Data, nil
+}
+
+// GetSecretValue gets and returns value for key of the referenced secret
+func (a *APISecretClient) GetSecretValue(ctx context.Context, sel xpv1.SecretKeySelector) ([]byte, error) {
+	d, err := a.GetSecretData(ctx, &sel.SecretReference)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot get secret data")
+	}
+	return d[sel.Key], err
+}
+
+// NewAPICallbacks returns a new APICallbacks.
+func NewAPICallbacks(m ctrl.Manager, of xpresource.ManagedKind) *APICallbacks {
+	nt := func() resource.Terraformed {
```

Comment:
I opted for this pattern so that we need to keep only the `metadata.name` of the resource rather than the whole object, which is different than the earlier implementation.